### ISSUE TITLE
Support `Optional` with null-safe and Elvis operators in SpEL expressions

### ISSUE THREAD

**Issue by spring-projects-issues** [spring-projects/spring-framework#20433]
**[Peter Luttrell](https://jira.spring.io/secure/ViewProfile.jspa?name=pluttrell)** opened **[SPR-15878](https://jira.spring.io/browse/SPR-15878?redirect=false)** and commented

This is a feature request to add support for Java 8 Optionals to the Spring Expression Language.

One use case that I just ran into is wanting to use `@PostAuthorize` on a method that returns an Optional in conjunctions with custom expressions. For example the following fails:

```java
@PostAuthorize("canAccessOrganization(returnObject.organiztionId)")
public Optional<Person> getPerson(String personId){
    ...
}
```

In this case, if the returned reference isn't present, that `@PostAuthorize` would allow the response, which should be Optional.empty(). If it is present, then it'd be dereferenced into the returnObject, so we'd have direct access to its fields for use in the expression.

---

4 votes, 6 watchers

**Comment by wimdeblauwe** [comment]
Another use case is with `@PreAuthorize`. Assume a `Task` has a reference to a `User` and there is a `TaskServiceImpl` class with:

```
    Optional<Task> getTask(int taskId);
```

The controller could check if the task is linked to the authorized user like this:

```
    @DeleteMapping("/{taskId}")
    @PreAuthorize("@taskServiceImpl.getTask(#taskId).orElse(null)?.user.id == #userDetails.id")
    public String destroy(@AuthenticationPrincipal ApplicationUserDetails userDetails,
                          @PathVariable("taskId") Integer taskId) {
        return "redirect:/tasks";
    }
```

The `orElse(null)` is perfect for this, not sure if anything else is needed, but just wanted to let you know about this use case.

**Comment by sanatik** [comment]
My use case would be with `@Cacheable`
```
@Cacheable(unless = "#result.isEmpty()")
public Optional<User> getUserById(final String userId);
```
Currently, I do not even know how to make it work. If I use `#result.isEmpty()` I get an exception `SpelEvaluationException: EL1004E: Method call: Method isEmpty() cannot be found on type com.example.User`.
If I use `#result == null`, the SPeL is being ignored (I guess because `Optional` is never `null`).

**Comment by dkfellows** [comment]
At the very least, `@PreFilter` and `@PostFilter` ought to work with `Optional` as if they were collections containing at most one object. If the filter sees a non-empty optional but doesn't want to let the contained object through, that becomes an empty optional; the _obvious_ semantics.

**Comment by OrangeDog** [comment]
This also applies to Thymeleaf templating, which uses SpEL in the default Boot configuration.

**Comment by cheatmenot** [comment]
+1

**Comment by lauroschuck** [comment]
This is causing me so much trouble now! I created a custom company lib that would benefit a lot from SpEL, and this is essentially making it useless because a lot of our codebase has a bias towards using Optional.

@jhoeller Why declined? Why not planned? This sounds like such a resonable request, so many people here voicing their support, and probably very natural do implement!

**Comment by drekbour** [comment]
Indeed. Lack of this caused us to drop SpEL long ago as the basis for something.

**Comment by sbrannen** [comment]
After further consideration, we have decided to introduce first-class support for `Optional` in SpEL expressions in Spring Framework 7.0. This aligns nicely with our overall null-safety efforts with JSpecify annotations and Kotlin improvements.

The discussions in this issue have presented two separate use cases.

1. transparent unwrapping of `Optional` for property and field access
2. transparent unwrapping on `Optional` when invoking a constructor, method, or function

After experimentation, I have determined that supporting `Optional` with the null-safe and Elvis operators would address use case `#1`.

To address use case `#2`, I have created #34544 to introduce an `OptionalToObjectConverter`.

**Comment by sbrannen** [comment]
> After experimentation, I have determined that supporting `Optional` with the null-safe and Elvis operators would address use case `#1`.

A proof of concept can be viewed in the following feature branch.

https://github.com/spring-projects/spring-framework/compare/main...sbrannen:spring-framework:issues/gh-20433-spel-null-safe-elvis-optional

That allows the original use case:

```java
@PostAuthorize("canAccessOrganization(returnObject.organizationId)")
public Optional<Person> getPerson(String personId){
    // ...
}
```

... to be replaced with:

```java
@PostAuthorize("canAccessOrganization(returnObject?.organizationId)")
public Optional<Person> getPerson(String personId){
    // ...
}
```

If `returnObject` is either `null` or an _empty_ `Optional`, `null` would be passed to the `canAccessOrganization()` invocation.

If `returnObject` is not `null` and not an `Optional`, `returnObject.organizationId` would be passed to the `canAccessOrganization()` invocation.

If `returnObject` is a non-null `Optional`, `returnObject.get().organizationId` would be passed to the `canAccessOrganization()` invocation, transparently _unwrapping_ the `Optional`.

@wimdeblauwe, you should then be able to rewrite your example as follows.

```java
    @DeleteMapping("/{taskId}")
    @PreAuthorize("@taskServiceImpl.getTask(#taskId)?.user.id == #userDetails.id")
    public String destroy(@AuthenticationPrincipal ApplicationUserDetails userDetails,
                          @PathVariable("taskId") Integer taskId) {
        return "redirect:/tasks";
    }
```

**Comment by OrangeDog** [comment]
> `returnObject.organizationId` would be passed to the `canAccessOrganization()` invocation.

Do you mean `returnObject.get().organizationId`? Otherwise it still won't work.

**Comment by sbrannen** [comment]
> > `returnObject.organizationId` would be passed to the `canAccessOrganization()` invocation.
> 
> Do you mean `returnObject.get().organizationId`?

No, in that particular example `returnObject` is "not an `Optional`".

> > `@taskServiceImpl.getTask(#taskId)?.user.id`
> 
> Do you mean `@taskServiceImpl.getTask(#taskId)?.user?.id` ?

Yes, that would indeed be better. I was merely modifying the original example to make use of the proposed new feature. But I'll update that to use null-safe navigation throughout the compound expression.

**Comment by sbrannen** [comment]
This support will be available in the upcoming Spring Framework 7.0 M3 release.

For details, check out the updated sections of the reference manual.

- [Null-safe Operations on `Optional`](https://docs.spring.io/spring-framework/reference/7.0-SNAPSHOT/core/expressions/language-ref/operator-safe-navigation.html#expressions-operator-safe-navigation-optional)
- [`Optional` Support with the Elvis Operator](https://docs.spring.io/spring-framework/reference/7.0-SNAPSHOT/core/expressions/language-ref/operator-elvis.html)

Feedback is welcome! 😎

**Comment by sbrannen** [comment]
> The discussions in this issue have presented two separate use cases.
> 
>     1. transparent unwrapping of `Optional` for property and field access
> 
>     2. transparent unwrapping on `Optional` when invoking a constructor, method, or function
> 
> 
> To address use case `#2`, I have created [#34544](https://github.com/spring-projects/spring-framework/issues/34544) to introduce an `OptionalToObjectConverter`.

FYI: I introduced an `OptionalToObjectConverter` in 8379ac772af44ac56b5cd9752822a5a6f3c770de, which is registered by default with the `DefaultConversionService`.

@stoussaint, it would be great if you could give it a try to see if it addresses your needs. You can experiment with in 7.0 [snapshots](https://github.com/spring-projects/spring-framework/wiki/Spring-Framework-Artifacts#snapshot-repository-configuration).
